File: dp/minimum_deletions.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends
// User function template for C++

class Solution {
  int dp[1001][1001];
  int go(const string &s, int i, int j) {
    if (i >= j)
      return 0;
    if (dp[i][j] != -1)
      return dp[i][j];
    int &ans = dp[i][j];
    if (s[i] == s[j]) {
      ans = go(s, i + 1, j - 1);
    } else {
      ans = go(s, i + 1, j) + 1;
      ans = min(ans, go(s, i, j - 1) + 1);
    }
    return ans;
  }

public:
  int minimumNumberOfDeletions(string S) {
    // code here
    memset(dp, -1, sizeof(dp));
    return go(S, 0, S.size() - 1);
  }
};

//{ Driver Code Starts.
int main() {
  int t;
  cin >> t;
  while (t--) {
    string S;
    cin >> S;
    Solution obj;
    cout << obj.minimumNumberOfDeletions(S) << endl;
  }
  return 0;
}
// } Driver Code Ends

File: dp/minimum_deletions.py
# User function Template for python3


class Solution:
    def minimumNumberOfDeletions(self, S):
        # code here
        n = len(S)
        dp = [[n for _ in range(n)] for _ in range(n)]
        for i in range(n):
            dp[i][i] = 0
        for i in range(n - 1):
            dp[i][i + 1] = 0 if S[i] == S[i + 1] else 1

        for k in range(2, n):
            for i in range(n - k):
                j = i + k
                if S[i] == S[j]:
                    dp[i][j] = dp[i + 1][j - 1]
                else:
                    dp[i][j] = min(dp[i + 1][j], dp[i][j - 1]) + 1

        return dp[0][n - 1]


# {
# Driver Code Starts
# Initial Template for Python 3

if __name__ == "__main__":
    t = int(input())
    for _ in range(t):
        S = input()

        ob = Solution()
        print(ob.minimumNumberOfDeletions(S))
# } Driver Code Ends
